### PR TITLE
Fire Support Menu for GM's

### PR DESCRIPTION
**Pull request by sunofang**
# About the pull request
Adds a fire support menu to allow game masters to call in various types of ordnance, and make it have the same noises/effects as a normal shot for extra soul.

# Explain why it's good for the game
uhh- good for GM's to be able to explode stuff at will.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>
image

</details>

# Changelog

:cl:
add: Due to budget increases, Close Air Support now has ammo.
admin: Added a fire support menu for GM's to use.
/:cl:

### CODE REVIEW

**Comment by bearrrrrrrr** [comment]
i love this so much it's unreal

will give feedback after dis op

**Comment by bearrrrrrrr** [comment]
no complaints

seems as if one gm swapping on the panel changes it for all gm's, though? not a big deal if it's not fixable

**Comment by sunofang** [comment]
Would need to make the vars client specific so it'd need alot of work for little to no gain.

**Comment by harryob** [comment]
address ci failure here https://github.com/cmss13-devs/cmss13-pve/actions/runs/9814119437/job/27101290362?pr=321#step:7:28

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+		addtimer(VARSET_CALLBACK(src, soundCooldown, FALSE), 10 SECONDS)
+
+/datum/fire_support_menu/proc/HandleDropshipOrdnance(turf/target_turf, obj/structure/ship_ammo/ammo)
+	addtimer(CALLBACK(src, PROC_REF(HandleDropshipSound), target_turf), 0.5 SECONDS)
+	if(!istype(ammo, /obj/structure/ship_ammo/heavygun/))
+		addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(playsound), target_turf, ammo.warning_sound, ammo.warning_sound_volume, 1, 15), 1.5 SECONDS)
+	addtimer(CALLBACK(ammo, TYPE_PROC_REF(/obj/structure/ship_ammo, detonate_on), target_turf), 2.5 SECONDS)
+	QDEL_IN(ammo, 5 SECONDS)
+
+/datum/fire_support_menu/proc/HandleOrbitalOrdnance(turf/target_turf, obj/structure/ob_ammo/warhead/ammo)
```
Comment (harryob) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:255]: procs should be snake case

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+#define FIRE_SUPPORT_CLICK_INTERCEPT_ACTION "fire_support_click_intercept_action"
+
+//Various ordnance selections
+#define ORDNANCE_OPTIONS list("Banshee Missile", "Harpoon Missile", "Keeper Missile", "Napalm Missile", "Thermobaric Missile", "Widowmaker Missile", "Laser", "Minirocket", "Incendiary Minirocket",  "Sentry Drop", "GAU-21", "Heavy GAU-21", "High Explosive", "Incendiary", "Cluster", "High Explosive", "Incendiary", "Fragmentation", "Flare")
+#define MISSILE_ORDNANCE list("Banshee Missile", "Harpoon Missile", "Keeper Missile", "Napalm Missile", "Thermobaric Missile", "Widowmaker Missile")
+#define ORBITAL_ORDNANCE list("High Explosive OB", "Incendiary OB", "Cluster OB")
+#define MORTAR_ORDNANCE list("High Explosive Shell", "Incendiary Shell", "Fragmentation Shell", "Flare Shell")
+#define MISC_ORDNANCE list("Laser", "Minirocket", "Incendiary Minirocket",  "Sentry Drop", "GAU-21", "Heavy GAU-21")
+
+client/proc/toggle_fire_support_menu()
```
Comment (harryob) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:11]:
```suggestion
/client/proc/toggle_fire_support_menu()
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	set category = "Game Master.Extras"
+	if(!check_rights(R_ADMIN))
+		return
+	GLOB.fire_support_menu.tgui_interact(mob)
+
+/datum/fire_support_menu
+	var/fire_support_click_intercept = FALSE
+	var/selected_ordnance = "Banshee Missile"
+	var/soundCooldown = FALSE
+	var/obj/structure/mortar/abstract_mortar = new()
```
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:22]: QDEL_NULL it on destroy

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+	set name = "Fire Support Menu"
+	set category = "Game Master.Extras"
+	if(!check_rights(R_ADMIN))
+		return
+	GLOB.fire_support_menu.tgui_interact(mob)
+
+/datum/fire_support_menu
+	var/fire_support_click_intercept = FALSE
+	var/selected_ordnance = "Banshee Missile"
+	var/soundCooldown = FALSE
```
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:21]: snake case

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+/datum/fire_support_menu/ui_data(mob/user)
+	. = ..()
+	var/list/data = list()
+
+	data["ordnance_options"] = ORDNANCE_OPTIONS
+
+	data["missile_ordnance_options"] = MISSILE_ORDNANCE
+	data["orbital_ordnance_options"] = ORBITAL_ORDNANCE
+	data["mortar_ordnance_options"] = MORTAR_ORDNANCE
+	data["misc_ordnance_options"] = MISC_ORDNANCE
```
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:40]: these can be static ui data

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm (hunk 1 above)
Comment (sunofang) [reply code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:255]: Good catch.

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    return UI_INTERACTIVE
+
+/datum/fire_support_menu/ui_close(mob/user)
+	var/client/user_client = user.client
+	if(user_client?.click_intercept == src)
+		user_client.click_intercept = null
+
+	fire_support_click_intercept = FALSE
+
+/datum/fire_support_menu/proc/InterceptClickOn(mob/user, params, atom/object)
```
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:67]: code doc these procs please

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm (hunk 3 above)
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:22]: code doc

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 7]
```diff
@@ -239,22 +242,29 @@ GLOBAL_DATUM_INIT(fire_support_menu, /datum/fire_support_menu, new)
...
 		addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(playsound), target_turf, ammo.warning_sound, ammo.warning_sound_volume, 1, 15), 1.5 SECONDS)
 	addtimer(CALLBACK(ammo, TYPE_PROC_REF(/obj/structure/ship_ammo, detonate_on), target_turf), 2.5 SECONDS)
 	QDEL_IN(ammo, 5 SECONDS)
 
-/datum/fire_support_menu/proc/HandleOrbitalOrdnance(turf/target_turf, obj/structure/ob_ammo/warhead/ammo)
+/datum/fire_support_menu/proc/handle_orbital_ordnance(turf/target_turf, obj/structure/ob_ammo/warhead/ammo)
 	ammo.warhead_impact(target_turf)
 
+///Deletes the mortar when the menu is closed so we dont make a thousand of them.
+/datum/fire_support_menu/Destroy(force, ...)
```
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:264]: stylistic nit: Destroy() overrides should be directly below Initialize() or the declaration of the type itself

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 8]
```diff
@@ -239,22 +242,29 @@ GLOBAL_DATUM_INIT(fire_support_menu, /datum/fire_support_menu, new)
...
+	QDEL_NULL(abstract_mortar)
+	. = ..()
```
Comment (Zonespace27) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:266]:
```suggestion
	return ..()
```

**Review by Zonespace27 — DISMISSED** [review]

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+		ui = new(user, src, "GameMasterFireSupportMenu", "Fire Support Menu")
+		ui.open()
+	user.client?.click_intercept = src
+
+///Input all the options for the ordnance panel. Only fires once, as the available ammo types shouldnt change. And if they do, something's very wrong
+/datum/fire_support_menu/ui_static_data(mob/user)
+	. = ..()
+	var/list/data = list()
+
+		data["ordnance_options"] = ORDNANCE_OPTIONS
```
Comment (Doubleumc) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:43]: indentation

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+		data["missile_ordnance_options"] = MISSILE_ORDNANCE
+		data["orbital_ordnance_options"] = ORBITAL_ORDNANCE
+		data["mortar_ordnance_options"] = MORTAR_ORDNANCE
+		data["misc_ordnance_options"] = MISC_ORDNANCE
```
Comment (Doubleumc) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:48]:
```suggestion
	data["ordnance_options"] = ORDNANCE_OPTIONS

	data["missile_ordnance_options"] = MISSILE_ORDNANCE
	data["orbital_ordnance_options"] = ORBITAL_ORDNANCE
	data["mortar_ordnance_options"] = MORTAR_ORDNANCE
	data["misc_ordnance_options"] = MISC_ORDNANCE
```

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+/datum/fire_support_menu/ui_status(mob/user, datum/ui_state/state)
+	return UI_INTERACTIVE
+
+/datum/fire_support_menu/ui_close(mob/user)
+	var/client/user_client = user.client
+	if(user_client?.click_intercept == src)
+		user_client.click_intercept = null
+
+	fire_support_click_intercept = FALSE
```
Comment (Doubleumc) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:80]:
```suggestion
	fire_support_click_intercept = FALSE
	qdel(src)
```

File: code/modules/admin/game_master/extra_buttons/fire_support_menu.dm [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+/datum/fire_support_menu/Destroy(force, ...)
+	QDEL_NULL(abstract_mortar)
+	return ..()
+
+/datum/fire_support_menu/tgui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "GameMasterFireSupportMenu", "Fire Support Menu")
+		ui.open()
+	user.client?.click_intercept = src
```
Comment (Doubleumc) [inline code/modules/admin/game_master/extra_buttons/fire_support_menu.dm:36]:
```suggestion
```

**Review by Doubleumc — CHANGES_REQUESTED** [review]
These changes will allow each GM to use the panel independently of what other GMs are doing.

Beyond that, looks good to me.

**Comment by sunofang** [comment]
One little thing. The fire support menu verb doesnt disappear when you de-admin. Any idea how to fix that? I'd appreciate if someone told me over discord.

**Comment by Doubleumc** [comment]
> One little thing. The fire support menu verb doesnt disappear when you de-admin. Any idea how to fix that? I'd appreciate if someone told me over discord.

The reason it appears in the first place is you've added the verb to `/client/proc/add_admin_verbs()`.

To have it disappear, you'll want to also add it to `/client/proc/remove_admin_verbs()`.

**Review by Doubleumc — DISMISSED** [review]

**Pull request merged**